src/spec.rs

```diff
@@ -0,0 +1,365 @@
+use std::ops::Index;
+
+use crate::{grain::Grain, matrix::Matrix};
+use pairing::arithmetic::FieldExt;
+
+/// `State` is structure `T` sized field elements that are subjected to
+/// permutation
+#[derive(Clone, Debug, PartialEq)]
+pub struct State<F: FieldExt, const T: usize>(pub(crate) [F; T]);
+
+impl<F: FieldExt, const T: usize> Default for State<F, T> {
+    /// The capacity value is 2**64 + (o − 1) where o the output length.
+    fn default() -> Self {
+        let mut state = [F::zero(); T];
+        // TODO/FIX: should capacity value placed in 0th element or (t-1)th element
+        state[0] = F::from_u128(1 << 64);
+        State(state)
+    }
+}
+
+impl<F: FieldExt, const T: usize> State<F, T> {
+    /// Applies sbox for all elements of the state.
+    /// Only supports `alpha = 5` sbox case.
+    pub(crate) fn sbox_full(&mut self) {
+        for e in self.0.iter_mut() {
+            let tmp = e.mul(*e);
+            e.mul_assign(tmp);
+            e.mul_assign(tmp);
+        }
+    }
+
+    /// Partial round sbox applies sbox to the first element of the state.
+    /// Only supports `alpha = 5` sbox case
+    pub(crate) fn sbox_part(&mut self) {
+        let tmp = self.0[0].mul(self.0[0]);
+        self.0[0].mul_assign(tmp);
+        self.0[0].mul_assign(tmp);
+    }
+
+    /// Adds constants to all elements of the state
+    pub(crate) fn add_constants(&mut self, constants: &[F; T]) {
+        for (e, constant) in self.0.iter_mut().zip(constants.iter()) {
+            e.add_assign(constant)
+        }
+    }
+
+    /// Only adds a constant to the first element of the state.It is used with
+    /// optimized rounds constants where only single element is added in
+    /// each partial round
+    pub(crate) fn add_constant(&mut self, constant: &F) {
+        self.0[0].add_assign(constant)
+    }
+
+    /// Copies elements of the state
+    pub(crate) fn words(&self) -> [F; T] {
+        self.0
+    }
+
+    /// Second element of the state is the result
+    pub(crate) fn result(&self) -> F {
+        self.0[1]
+    }
+}
+
+/// `Spec` holds construction parameters as well as constants that are used in
+/// permutation step. Constants are planned to be hardcoded once transcript
+/// design matures. Number of partial rounds can be deriven from number of
+/// constants.
+#[derive(Debug, Clone)]
+pub struct Spec<F: FieldExt, const T: usize, const RATE: usize> {
+    pub r_f: usize,
+    pub mds_matrices: MDSMatrices<F, T, RATE>,
+    pub constants: OptimizedConstants<F, T>,
```

Comment:
Transcript circuit requires these constants so easiest way to get them is exposing the fields but there might be more elegant way?

Reply:
Hmmm I see. 

Well I would make getter methods following the rust API standards.
Since otherways looks like they're public because they're meant to be mutated or whatever by the end user when it's not the case.